### PR TITLE
fix: ensure empty withdrawals after cancun

### PR DESCRIPTION
**Pull request by buddh0**
### Description

fix: ensure empty withdrawals after cancun

### Rationale

tell us why we need these changes...

### Example

add an example CLI or API response...

### Changes

Notable changes: 
* add each change in a bullet point here
* ...

### CODE REVIEW

File: core/block_validator.go [hunk 1]
```diff
@@ -96,15 +96,15 @@ func (v *BlockValidator) ValidateBody(block *types.Block) error {
 		},
 		func() error {
 			// Withdrawals are present after the Shanghai fork.
-			if !header.EmptyWithdrawalsHash() {
```
Comment (buddh0) [inline core/block_validator.go:99]: rollback and align with geth

File: core/block_validator.go [hunk 2]
```diff
@@ -96,15 +96,15 @@ func (v *BlockValidator) ValidateBody(block *types.Block) error {
...
+			if header.WithdrawalsHash != nil {
 				// Withdrawals list must be present in body after Shanghai.
 				if block.Withdrawals() == nil {
 					return errors.New("missing withdrawals in block body")
 				}
 				if hash := types.DeriveSha(block.Withdrawals(), trie.NewStackTrie(nil)); hash != *header.WithdrawalsHash {
 					return fmt.Errorf("withdrawals root hash mismatch (header value %x, calculated %x)", *header.WithdrawalsHash, hash)
 				}
-			} else if len(block.Withdrawals()) != 0 { // Withdrawals turn into empty from nil when BlockBody has Sidecars
```
Comment (buddh0) [inline core/block_validator.go:107]: rollback and align with geth

File: core/types/block.go [hunk 3]
```diff
@@ -195,7 +195,7 @@ func (h *Header) EmptyReceipts() bool {
 
 // EmptyWithdrawalsHash returns true if there are no WithdrawalsHash for this header/block.
 func (h *Header) EmptyWithdrawalsHash() bool {
-	return h.WithdrawalsHash == nil || *h.WithdrawalsHash == EmptyWithdrawalsHash
```
Comment (buddh0) [inline core/types/block.go:198]: make  it equivalent to other Emptyxxx function, such as EmptyReceipts

File: eth/downloader/queue.go [hunk 4]
```diff
@@ -81,7 +81,7 @@ func newFetchResult(header *types.Header, fastSync bool, pid string) *fetchResul
 	}
 	if !header.EmptyBody() {
 		item.pending.Store(item.pending.Load() | (1 << bodyType))
-	} else if !header.EmptyWithdrawalsHash() {
```
Comment (buddh0) [inline eth/downloader/queue.go:84]: rollback and align with geth

File: eth/downloader/queue.go [hunk 5]
```diff
@@ -788,9 +788,9 @@ func (q *queue) DeliverBodies(id string, txLists [][]*types.Transaction, txListH
 		if uncleListHashes[index] != header.UncleHash {
 			return errInvalidBody
 		}
-		if header.EmptyWithdrawalsHash() {
```
Comment (buddh0) [inline eth/downloader/queue.go:791]: rollback and align with geth

File: eth/downloader/queue.go [hunk 6]
```diff
@@ -788,9 +788,9 @@ func (q *queue) DeliverBodies(id string, txLists [][]*types.Transaction, txListH
...
+		if header.WithdrawalsHash == nil {
 			// nil hash means that withdrawals should not be present in body
-			if len(withdrawalLists[index]) != 0 {
```
Comment (buddh0) [inline eth/downloader/queue.go:793]: rollback and align with geth

File: graphql/graphql.go [hunk 7]
```diff
@@ -1053,7 +1053,7 @@ func (b *Block) Withdrawals(ctx context.Context) (*[]*Withdrawal, error) {
 		return nil, err
 	}
 	// Pre-shanghai blocks
-	if block.Header().EmptyWithdrawalsHash() {
```
Comment (buddh0) [inline graphql/graphql.go:1056]: rollback and align with geth

File: miner/worker.go [hunk 8]
```diff
@@ -1367,6 +1367,10 @@ func (w *worker) commit(env *environment, interval func(), update bool, start ti
 		// env.receipts = receipts
 		finalizeBlockTimer.UpdateSince(finalizeStart)
 
+		if block.Header().EmptyWithdrawalsHash() {
```
Comment (buddh0) [inline miner/worker.go:1370]: set empty withdraws

File: eth/fetcher/block_fetcher.go [hunk 9]
```diff
@@ -723,6 +723,9 @@ func (f *BlockFetcher) loop() {
 						matched = true
 						if f.getBlock(hash) == nil {
 							block := types.NewBlockWithHeader(announce.header).WithBody(task.transactions[i], task.uncles[i])
+							if block.Header().EmptyWithdrawalsHash() {
```
Comment (buddh0) [inline eth/fetcher/block_fetcher.go:726]: set empty withdraws at fetcher side

File: core/types/block.go [hunk 10]
```diff
@@ -195,7 +195,7 @@ func (h *Header) EmptyReceipts() bool {
 
 // EmptyWithdrawalsHash returns true if there are no WithdrawalsHash for this header/block.
```
Comment (emailtovamos) [inline core/types/block.go:196]: Maybe we should change the description of this function to make it more clear.

File: eth/fetcher/block_fetcher.go (hunk 9 above)
Comment (emailtovamos) [reply eth/fetcher/block_fetcher.go:726]: Do we need similar changes in downloader here? 
https://github.com/bnb-chain/bsc/blob/38db9bf4e29b00e5675d4cd9297b9a979ecea7b6/eth/downloader/downloader.go#L1392

Comment (buddh0) [reply eth/fetcher/block_fetcher.go:726]: > Do we need similar changes in downloader here?
> 
> https://github.com/bnb-chain/bsc/blob/38db9bf4e29b00e5675d4cd9297b9a979ecea7b6/eth/downloader/downloader.go#L1392

esult.Withdrawals is already empty now

File: core/types/block.go (hunk 10 above)
Comment (buddh0) [reply core/types/block.go:196]: > Maybe we should change the description of this function to make it more clear.

ok

File: eth/fetcher/block_fetcher.go (hunk 9 above)
Comment (emailtovamos) [reply eth/fetcher/block_fetcher.go:726]: Ok. But under what condition the above if statement of fetcher will be true?

File: eth/protocols/eth/handlers.go [hunk 11]
```diff
@@ -367,7 +367,7 @@ func handleBlockBodies(backend Backend, msg Decoder, peer *Peer) error {
 			txsHashes[i] = types.DeriveSha(types.Transactions(body.Transactions), hasher)
 			uncleHashes[i] = types.CalcUncleHash(body.Uncles)
 			// Withdrawals may be not nil, but a empty value when Sidecars not empty
-			if len(body.Withdrawals) > 0 {
```
Comment (buddh0) [inline eth/protocols/eth/handlers.go:370]: rollback and align with geth

**Review by emailtovamos — APPROVED** [review]

**Comment by zzzckck** [comment]
for block header/body payload, pls refer: https://forum.bnbchain.org/t/faq-about-eip-4844-on-bsc/2620#q11-block-headerbody-payload-change-11

**Review by zzzckck — APPROVED** [review]

**Pull request merged**